File: Longest Increasing Subsequence - GFG/longest-increasing-subsequence.java
//{ Driver Code Starts
import java.io.*;
import java.util.*; 

class GFG{
    public static void main(String args[]) throws IOException { 
        
        //taking input using Scanner class
        Scanner sc = new Scanner(System.in);
        
        //taking total testcases
        int t = sc.nextInt();
        while(t > 0){
            
            //taking size of array
            int n = sc.nextInt();
            int array[] = new int[n];
            
            //inserting elements in the array
            for (int i = 0; i < n; ++i)
            {
                array[i] = sc.nextInt();
            }
            
            //creating an object of class Solution
            Solution ob = new Solution();
            
            //calling longestSubsequence() method of class
            //Solution
            System.out.println(ob.longestSubsequence(n,array));
            t--;
        }
    } 
} 
// } Driver Code Ends

class Solution 
{
    //Function to find length of longest increasing subsequence.
    static int longestSubsequence(int n, int a[])
    {
        // code here
        ArrayList<Integer> l = new ArrayList<>();
        l.add(a[0]);
        for(int i=1;i<n;i++)
        {
            int pos=Collections.binarySearch(l,a[i]);
            if(pos<0){
            pos=Math.abs(pos)-1;
            }
            if(pos<l.size())
            l.set(pos,a[i]);
            else 
            l.add(a[i]);
        }
        return l.size();
    }
} 

File: Longest Increasing Subsequence - GFG/longest-increasing-subsequence.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends


class Solution
{
    public:
    //Function to find length of longest increasing subsequence.
    int longestSubsequence(int n, int a[])
    {
       // your code here
       vector<int> temp;
       temp.push_back(a[0]);
       int i=1;
       while(i<n){
           if(a[i]>temp[temp.size()-1])
           temp.push_back(a[i]);
           else
           {
               int x=lower_bound(temp.begin(),temp.end(),a[i])-temp.begin();
               temp[x]=a[i];
           }
           i++;
       }
       return temp.size();
    }
};

//{ Driver Code Starts.
int main()
{
    //taking total testcases
    int t,n;
    cin>>t;
    while(t--)
    {
        //taking size of array
        cin>>n;
        int a[n];
        
        //inserting elements to the array
        for(int i=0;i<n;i++)
            cin>>a[i];
        Solution ob;
        //calling method longestSubsequence()
        cout << ob.longestSubsequence(n, a) << endl;
    }
}

// } Driver Code Ends